File: Dynamic Programming/GFG-Must Do Questions/knapsack.cpp
#include <iostream>
#include<vector>
using namespace std;

int knapsack(vector<int> &weights,vector<int> &val,int capacity,int i,vector<vector<int> > &dp){
    if(i==weights.size()){
        return 0;
    }
    
    if(dp[i][capacity]!=-1){
        return dp[i][capacity];
    }
    
    int op1=0;
    if(capacity>=weights[i]){
        op1=val[i]+knapsack(weights,val,capacity-weights[i],i+1,dp);
    }
    int op2=knapsack(weights,val,capacity,i+1,dp);
    dp[i][capacity]=max(op1,op2);
    return dp[i][capacity];
    
}

int main() {
	//code
	int t;
	cin>>t;
	for(int i=0;i<t;i++){
	    int n;
	    int c;
	    cin>>n>>c;
	    vector<int> values;
	    vector<int> weights;
	    for(int j=0;j<n;j++){
	        int temp;
	        cin>>temp;
	        values.push_back(temp);
	    }
	    for(int j=0;j<n;j++){
	        int temp;
	        cin>>temp;
	        weights.push_back(temp);
	    }
	    
	    vector<vector<int> > dp(n,vector<int> (c+1,-1));
	    cout<<knapsack(weights,values,c,0,dp)<<endl;
	}
	return 0;
}

File: Python practice/DP/knapsack.java
public class knapsack {
    public static void main(String[] args) {
        int[] wt = {1,3,4,5};
        int[] val = {1,4,5,7};
        int w = 7;
        int n = 4;
        System.out.println(knap(wt, val, w, n));

    }
    public static int knap(int[] wt, int[] val, int w, int n){
        if (n == 0 || w == 0)
            return 0;
        if (wt[n-1] <= w)
            return Math.max((val[n-1] + knap(wt, val, w-wt[n-1], n-1)), (knap(wt, val, w, n-1)));
        if (wt[n-1] > w)
            return knap(wt, val, w, n-1);
        return w;
    }
}
